cudf implementation dataframe protocol (#265)
Quansight-Labs/quansight-labs-site: posts/2021/10/interchange-dataframe-protocol-for-cudf.md

<br/>

<p align="center">
    <img
     alt="A composition of the 3 interfaces forming the dataframe interchange protocol: `_CuDFDataFrame` has 1 or more `_CuDFColumn` which in turn has 1 or more `_CuDFBuffer` "
     src="/images/2021/10/dataframe-api-cudf/protocol_interfaces.jpg">
    <i>Composition of the interchange dataframe protocol interfaces</i>
</p>
<br/>
               
So each library supporting the protocol should implement these 3 interfaces that can be described as follows:

- **DataFrame** mainly exposes different methods to access/select columns (by name, index) and knowing the number of rows.

- **Column** has methods to access the column data type, describe  valid/missing values, exposes different buffers (data, validity and offset),  chunks, ...etc.

- **Buffer** has methods to describe the contiguous block of memory of the column data i.e device (GPU, CPU, ...), memory address, size, etc...

For more details, please have a look at [python code of the protocol interfaces](https://github.com/data-apis/dataframe-api/blob/main/protocol/dataframe_protocol.py) and [design concepts](https://github.com/data-apis/dataframe-api/blob/main/protocol/design_requirements.md).

> ### What is expected from cudf implementation

Author: You're totally right. Thank you.

    0  1000      0  <NA>   True        1000
    1     2    128   2.5   <NA>           2
    2   300    255  <NA>  False         300
    3  <NA>     25  10.0   True        <NA> 

    <class 'cudf.core.dataframe.DataFrame'>
    RangeIndex: 4 entries, 0 to 3
    Data columns (total 5 columns):
     #   Column       Non-Null Count  Dtype
    ---  ------       --------------  -----
     0   int          3 non-null      int64
     1   uint8        4 non-null      uint8
     2   float        2 non-null      float64
     3   bool         3 non-null      bool
     4   categorical  3 non-null      category
    dtypes: bool(1), category(1), float64(1), int64(1), uint8(1)
    memory usage: 356.0 bytes

Now, we create the interchange protocol dataframe object and check that basic information like number of rows, column names and dtypes are accurate:
```python
dfo =  df.__dataframe__()
print(f'{dfo}: {dfo.num_rows()} rows')
print('Column name\t Non-Null Count\t\t\t\t    Dtype')
print('-----------\t --------------\t\t\t\t    -----')
for n, c in zip(dfo.column_names(), dfo.get_columns()): 
    print(f'{n}\t\t\t      {int(c.size - c.null_count)}\t\t\t{c.dtype}')
```
**output**:

>     <cudf.core.df_protocol._CuDFDataFrame object at 0x7ff90ac10ca0>: 4 rows

Author: Thank you for the spot .

Thus, a diversity of levels(experts, new comers, etc...) ensures an inclusive environment where everyone can find his way easily.

### Test Driven Development (TDD)
This process has been very helpful during this project. I've noticed my slowness on days where I've started developing features before writing any test. I kept going back and forth in the code base to ensure the coherence of different pieces of code written for that feature. However, when writing tests I felt the possibility to express all my expectations across different test cases then writing code for each one at a time. In this process, I felt like the tests pointed out next place on the code base where they might be something wrong. 

### Collaboration, speaking out is very helpful

> Sometimes when stumble upon a problem, just speaking out or sharing the problem to someone else opens your eyes to a possible solution. This happens to me countless when discussing with my mentor @Kshiteej and my colleague @Alenka whose project is very close to mine. Otherwise, external inputs combined with ours will definitely be better than ours alone. 

Reviewer: ```suggestion
Sometimes, when stumbling upon a problem, just speaking out or sharing the problem to someone else opens your eyes to a possible solution. This happened to me countless times when discussing with my mentor @Kshiteej and my colleague @Alenka whose project is very close to mine. Otherwise, external inputs combined with ours will definitely be better than ours alone. 
```

### Patience and Perseverence

> It is well known that configuring and installing drivers to work with GPU is not an easy task. During this internship I've been rebuilding the cudf library many times and encountered multiple issues. I came up with a recipe which is: 
> - seek help after around 40 minutes of debugging even if you have more ideas to try. 
> - When asking for help, do share what you've tried and other ideas to try. 

Reviewer: ```suggestion
It is well known that configuring and installing drivers to work with GPU is not an easy task. During this internship I've been rebuilding the cudf library many times and encountered multiple issues. I came up with a recipe which is: 

- seek help after around 40 minutes of debugging even if you have more ideas to try. 
- When asking for help, do share what you've tried and other ideas to try. 
```

<!--
.. title: Interchange dataframe protocol: cuDF implementation
.. slug: data-apis-cudf

> .. date: 2021-10-05 16:10:00 UTC+00:00

Reviewer 2: ```suggestion
.. date: 2021-10-21 09:00:00 UTC+00:00
```
@iskode @melissawm i've modified the publish date for this thursday. are there any major problems that would hinder this from publishing on this coming thursday?
